reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java

```diff
@@ -367,15 +378,21 @@ void drainAsync() {
 					while (e != r) {
 						if (cancelled) {
 							current = null;
-							Operators.onDiscardQueueWithClear(q, ctx, null);
+							currentKnownToBeFinite = false; //reset explicitly
```

Comment:
this block seems to be repeated 9 times, wdyt about extracting it into a method?

Reply:
Maybe not the whole block, but the `current = null; currentKnownToBeFinite = false;` part: it is very important these two are reset together, and the code is short enough that it can be inlined. I'll extract that instead (19 duplicates).